lib/Mount/FolderStorageManager.php

```diff
@@ -54,7 +57,72 @@ public function getRootAndStorageIdForFolder(int $folderId): array {
 	/**
 	 * @param 'files'|'trash'|'versions' $type
 	 */
-	public function getBaseStorageForFolder(int $folderId, ?FolderDefinition $folder = null, ?IUser $user = null, bool $inShare = false, string $type = 'files'): IStorage {
+	public function getBaseStorageForFolder(
+		int $folderId,
+		bool $separateStorage,
+		?FolderDefinition $folder = null,
+		?IUser $user = null,
+		bool $inShare = false,
+		string $type = 'files',
+	): IStorage {
+		if ($separateStorage) {
+			return $this->getBaseStorageForFolderSeparateStorageLocal($folderId, $folder, $user, $inShare, $type);
+		} else {
+			return $this->getBaseStorageForFolderRootJail($folderId, $folder, $user, $inShare, $type);
+		}
+	}
+
+	/**
+	 * @param 'files'|'trash'|'versions' $type
+	 */
+	public function getBaseStorageForFolderSeparateStorageLocal(
+		int $folderId,
+		?FolderDefinition $folder = null,
+		?IUser $user = null,
+		bool $inShare = false,
+		string $type = 'files',
+	): IStorage {
+		$dataDirectory = $this->config->getSystemValue('datadirectory');
+		$rootPath = $dataDirectory . '/__groupfolders/' . $folderId;
+		$init = !is_dir($rootPath);
+		if ($init) {
+			mkdir($rootPath, 0777, true);
+			mkdir($rootPath . '/files');
+			mkdir($rootPath . '/trash');
+			mkdir($rootPath . '/versions');
+		}
+
+		$storage = new Local([
+			'datadir' => $rootPath,
+		]);
+
+		if ($init) {
+			$storage->getScanner()->scan('');
+		}
```

Comment:
Why not use `$storage->mkdir()` to create these folders?

Reply:
I want to do "raw" `mkdir` anyway because of the recursive, so might as well do that for the rest, `$storage->mkdir` has no advantage over raw mkdir here.